target/ppc/int_helper.c

```diff
@@ -1771,6 +1772,47 @@ void helper_xxinsertw(CPUPPCState *env, ppc_vsr_t *xt,
     *xt = t;
 }
 
+void helper_XXEVAL(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b, ppc_avr_t *c,
+                   uint32_t desc)
+{
+    /*
+     * Instead of processing imm bit-by-bit, we'll skip the computation of
+     * conjunctions whose corresponding bit is unset.
+     */
+    int bit, imm = simd_data(desc);
+    Int128 conj, disj = int128_zero();
+
+    /* Iterate over set bits from the least to the most significant bit */
+    while (imm) {
+        /*
+         * Get the next bit to be processed with ctz64. Invert the result of
+         * ctz64 to match the indexing used by PowerISA.
+         */
+        bit = 7 - ctz64(imm);
+        if (bit & 0x4) {
+            conj = a->s128;
+        } else {
+            conj = int128_not(a->s128);
+        }
+        if (bit & 0x2) {
+            conj = int128_and(conj, b->s128);
+        } else {
+            conj = int128_and(conj, int128_not(b->s128));
+        }
+        if (bit & 0x1) {
+            conj = int128_and(conj, c->s128);
+        } else {
+            conj = int128_and(conj, int128_not(c->s128));
+        }
+        disj = int128_or(disj, conj);
+
+        /* Unset the least significant bit that is set */
+        imm &= imm - 1;
+    }
+
+    t->s128 = disj;
+}
```

Comment:
I like how you optimized for not doing a loop over the number of bits in a/b/c, and iterated over imm instead.

A suggestion to optimize it further is to get rid of ctz64, especially in platforms that don't support it natively, with something like:
```
    int i = imm;
    int i_next;
    int masked_bit;

    // ...

    while (i) {
       i_next = i & (i - 1);
       masked_bit = i ^ i_next;

       /* A */
       if (masked_bit & 0b00001111) {
           // ...
       } else {
           // ...
       }

       /* B */
       if (masked_bit & 0b00110011) {
           // ...
       } else {
           // ...
       }

       /* C */
       if (masked_bit & 0b01010101) {
           // ...
       } else {
           // ...
       }

       // ...

       i = i_next;
    }
```
The first ``int i = imm`` does not really apply to this helper_ version, where imm is already an int. But it does apply to the other 2 versions (gen_xxeval_i64 and gen_xxeval_vec).

Reply:
That's a nice trick to get rid of `ctz64` (that should've be `ctzl` to match the type), but I'm not sure it's worth the impact on readability. Most supported platforms can count trailing zeros with one instruction (`bsf`/`tzcnt` on x86, `cnttzw`/`cnttzd` on ppc, `dctz` on some mips) or two (`rbit`+`clz` on arm), and `ctz*` is already widely used in TCG code.

Reply:
It's up to you. I will note, however, that you're using ctz inside a loop, and that this is a very restricted scenario, where you're counting bits in a value that's just 8 bits long (not 32 or 64), and your use of the result is also simple enough (3 trivial tests to mask A, B and C).